crypto/evp/digest.c

```diff
@@ -1146,10 +1155,42 @@ static void evp_md_free(void *md)
 EVP_MD *EVP_MD_fetch(OSSL_LIB_CTX *ctx, const char *algorithm,
                      const char *properties)
 {
-    EVP_MD *md =
-        evp_generic_fetch(ctx, OSSL_OP_DIGEST, algorithm, properties,
-                          evp_md_from_algorithm, evp_md_up_ref, evp_md_free);
+    EVP_MD *md;
+#ifndef FIPS_MODULE
+    HT_VALUE *v = NULL;
```

Comment:
Many algorithms use EVP_MD_fetch() internally.. e.g. KDF and MAC, So these will be slower in the fips provider?

Reply:
Yeah, I'd love to find a way around this, but its currently somewhat problematic.  Because rcu stores per-thread data, the accessing of a lock requires registration via ossl_init_thread_start to register the exit handler for the thread.  Normally this is fine, but when running in the fips provider, its impossible to get the thread handler callback array, without passing around the fips provider context.  If you have an idea for how to manage that, I'd welcome the input, but so far i've come up empty

Reply:
The FIPS provider has an internal libctx from which it fetches everything and this only contains the FIPS provider.  I don't think the property query string is even meaningful here (although it can be passed through) -- there is only one implementation of any algorithm that can be fetched so the worst the query can do is to cause the fetch to fail.

I.e. the FIPS provider might be able to just look up by algorithm name from a single shared hash table that could be built at load time.

Then again, there also isn't a lot to search through since only the FIPS approved algorithms are there, so performance might be good enough.